cmd/tools/genvisitor/main.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+
+	// Import to populate the protoregistry
+	_ "go.temporal.io/api/workflowservice/v1"
+	_ "go.temporal.io/server/api/adminservice/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+)
+
+func main() {
+	debugFlag := flag.Bool("debug", false, "enable debug logs to stderr")
+	dumpTree := flag.Bool("dump-tree", false, "print the tree of matched paths in the type hierarchy to stderr")
+	flag.Parse()
+
+	var logger log.Logger
+	if *debugFlag {
+		logger = log.NewCLILogger()
+	} else {
+		logger = log.NewNoopLogger()
+	}
+
+	emitter := NewEmitter(logger, Gogo122Version)
+	emitter.SetFunctionSignature(
+		`func repairInvalidUTF8(vAny any) (ret bool, retErr error)`,
+	)
+	emitter.SetFunctionTrailer("return")
+	emitter.AddHandler(
+		// Match any type called "Failure"
+		func(vt VisitType, path VisitPath) bool {
+			// Match Failure types, but not Cause fields.
+			if vt.GoTypeName() != "Failure" {
+				logger.Debug("ignore non Failure field", tag.NewAnyTag("path", path.String()))
+				return false
+			}
+			if strings.Contains(path.String(), "/Cause") {
```

Comment:
this is so we only visit the top level "Failure" type? maybe add a comment here.

Reply:
Right. We can have the `repairInvalidUTF8InFailure` function handle the top-level Failure, and descend in to the cause field.